From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.15 october2005, patch 02-38

  Description: tape: Kernel panic when tape operation is aborted with signals
  Symptom:     Kernel panic When hitting <CTRL>C during a long running tape
               operation.
  Problem:     When a request is aborted due to a signal, we currently
               stop the request via csh, but we do not wait for the interrupt
               of csh in any case. We free the request structure and therefore
               when the interrupt for the csh operation is presented, the request
               object is no longer valid and an invalid callback pointer is used.
  Solution:    Wait until the interrupt for csh arrives and until
               wait_event_interruptible() does not return -ERESTARTSYS.
  Problem-ID:  -
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/drivers/s390/char/tape_core.c	4 Jan 2006 17:59:18 -0000	1.56.2.1
+++ linux-2.5/drivers/s390/char/tape_core.c	17 Mar 2006 13:14:44 -0000	1.56.2.2
@@ -1015,7 +1015,7 @@
 				wq,
 				(request->callback == NULL)
 			);
-		} while (rc != -ERESTARTSYS);
+		} while (rc == -ERESTARTSYS);
 
 		DBF_EVENT(3, "IO stopped on %08x\n", device->cdev_id);
 		rc = -ERESTARTSYS;

